How to build this?

<!-- Thanks for filing an issue! Before hitting the button, please answer these questions. It's helpful to search the existing GitHub issues first. It's likely that another user has already reported the issue you're facing, or it's a known issue that we're already aware of-->

**Is this a request for help?**:
YES
---

**Is this an ISSUE or FEATURE REQUEST?** (choose one):

---

**Which release version?**:
master
---

**Which component (CNI/IPAM/CNM/CNS)**:

---

**Which Operating System (Linux/Windows)**:

Centos 7 / Linux localhost 3.10.0-693.21.1.el7.x86_64 #1 SMP Wed Mar 7 19:03:37 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
---

<!--
If this is a ISSUE, please:
  - Fill in as much of the template below as you can.  If you leave out
    information, we can't help you as well.

If this is a FEATURE REQUEST, please:
  - Describe *in detail* the feature/behavior/change you'd like to see.

In both cases, be ready for followup questions, and please respond in a timely
manner.  If we can't reproduce a bug or think a feature already exists, we
might close your issue.  If we're wrong, PLEASE feel free to reopen it and
explain why.
-->

**Which Orchestrator and version (e.g. Kubernetes, Docker)**


**What happened**:
make all-binaries fails

**What you expected to happen**:

make all-binaries success

**How to reproduce it** (as minimally and precisely as possible):

```
$ git clone https://github.com/Azure/azure-container-networking
$ cd azure-container-networking
$ make all-binaries
```
**Anything else we need to know**:

```
[user@localhost ~]$ git clone https://github.com/Azure/azure-container-networking
Cloning into 'azure-container-networking'...
remote: Counting objects: 2482, done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 2482 (delta 0), reused 2 (delta 0), pack-reused 2476
Receiving objects: 100% (2482/2482), 4.33 MiB | 3.73 MiB/s, done.
Resolving deltas: 100% (1448/1448), done.
```

```
cd azure-container-networking/
```

```
[user@localhost azure-container-networking]$ make all-binaries
go build -v -o output/linux_amd64/cnm/azure-vnet-plugin -ldflags "-X main.version=v1.0.5-1-g89555ac -s -w" cnm/plugin/*.go
cnm/plugin/main.go:12:2: cannot find package "github.com/Azure/azure-container-networking/cnm/ipam" in any of:
        /usr/lib/golang/src/github.com/Azure/azure-container-networking/cnm/ipam (from $GOROOT)
        /home/user/go/src/github.com/Azure/azure-container-networking/cnm/ipam (from $GOPATH)
cnm/plugin/main.go:13:2: cannot find package "github.com/Azure/azure-container-networking/cnm/network" in any of:
        /usr/lib/golang/src/github.com/Azure/azure-container-networking/cnm/network (from $GOROOT)
        /home/user/go/src/github.com/Azure/azure-container-networking/cnm/network (from $GOPATH)
cnm/plugin/main.go:14:2: cannot find package "github.com/Azure/azure-container-networking/common" in any of:
        /usr/lib/golang/src/github.com/Azure/azure-container-networking/common (from $GOROOT)
        /home/user/go/src/github.com/Azure/azure-container-networking/common (from $GOPATH)
cnm/plugin/main.go:15:2: cannot find package "github.com/Azure/azure-container-networking/log" in any of:
        /usr/lib/golang/src/github.com/Azure/azure-container-networking/log (from $GOROOT)
        /home/user/go/src/github.com/Azure/azure-container-networking/log (from $GOPATH)
cnm/plugin/main.go:16:2: cannot find package "github.com/Azure/azure-container-networking/platform" in any of:
        /usr/lib/golang/src/github.com/Azure/azure-container-networking/platform (from $GOROOT)
        /home/user/go/src/github.com/Azure/azure-container-networking/platform (from $GOPATH)
cnm/plugin/main.go:17:2: cannot find package "github.com/Azure/azure-container-networking/store" in any of:
        /usr/lib/golang/src/github.com/Azure/azure-container-networking/store (from $GOROOT)
        /home/user/go/src/github.com/Azure/azure-container-networking/store (from $GOPATH)
make: *** [output/linux_amd64/cnm/azure-vnet-plugin] Error 1
```
```
[user@localhost azure-container-networking]$ go get github.com/Azure/azure-container-networking
package github.com/Azure/azure-container-networking: no Go files in /home/user/go/src/github.com/Azure/azure-container-networking
```

```
[user@localhost ]$ cd /home/user/go/src/github.com/Azure/azure-container-networking
make all-binaries
go build -v -o output/linux_amd64/cnm/azure-vnet-plugin -ldflags "-X main.version=v1.0.5-1-g89555ac -s -w" cnm/plugin/*.go
netlink/ip.go:9:2: cannot find package "golang.org/x/sys/unix" in any of:
        /usr/lib/golang/src/golang.org/x/sys/unix (from $GOROOT)
        /home/user/go/src/golang.org/x/sys/unix (from $GOPATH)
make: *** [output/linux_amd64/cnm/azure-vnet-plugin] Error 1
```

```
[user@localhost azure-container-networking]$ go get golang.org/x/sys/unix
make all-binaries
go build -v -o output/linux_amd64/cnm/azure-vnet-plugin -ldflags "-X main.version=v1.0.5-1-g89555ac -s -w" cnm/plugin/*.go
github.com/Azure/azure-container-networking/platform
github.com/Azure/azure-container-networking/log
github.com/Azure/azure-container-networking/store
github.com/Azure/azure-container-networking/common
github.com/Azure/azure-container-networking/cnm
github.com/Azure/azure-container-networking/ipam
github.com/Azure/azure-container-networking/cnm/ipam
github.com/Azure/azure-container-networking/ebtables
github.com/Azure/azure-container-networking/netlink
github.com/Azure/azure-container-networking/network/policy
github.com/Azure/azure-container-networking/network
github.com/Azure/azure-container-networking/cnm/network
command-line-arguments
chmod 0755 output/linux_amd64/cnm/azure-vnet-plugin
cd output/linux_amd64/cnm && tar -czvf azure-vnet-cnm-linux-amd64-v1.0.5-1-g89555ac.tgz azure-vnet-plugin
azure-vnet-plugin
chown 1000:1000 output/linux_amd64/cnm/azure-vnet-cnm-linux-amd64-v1.0.5-1-g89555ac.tgz
go build -v -o output/linux_amd64/cni/azure-vnet -ldflags "-X main.version=v1.0.5-1-g89555ac -s -w" cni/network/plugin/*.go
cni/plugin.go:16:2: cannot find package "github.com/containernetworking/cni/pkg/invoke" in any of:
        /usr/lib/golang/src/github.com/containernetworking/cni/pkg/invoke (from $GOROOT)
        /home/user/go/src/github.com/containernetworking/cni/pkg/invoke (from $GOPATH)
cni/cni.go:7:2: cannot find package "github.com/containernetworking/cni/pkg/skel" in any of:
        /usr/lib/golang/src/github.com/containernetworking/cni/pkg/skel (from $GOROOT)
        /home/user/go/src/github.com/containernetworking/cni/pkg/skel (from $GOPATH)
cni/internal.go:10:2: cannot find package "github.com/containernetworking/cni/pkg/types" in any of:
        /usr/lib/golang/src/github.com/containernetworking/cni/pkg/types (from $GOROOT)
        /home/user/go/src/github.com/containernetworking/cni/pkg/types (from $GOPATH)
cni/plugin.go:19:2: cannot find package "github.com/containernetworking/cni/pkg/types/current" in any of:
        /usr/lib/golang/src/github.com/containernetworking/cni/pkg/types/current (from $GOROOT)
        /home/user/go/src/github.com/containernetworking/cni/pkg/types/current (from $GOPATH)
cni/plugin.go:20:2: cannot find package "github.com/containernetworking/cni/pkg/version" in any of:
        /usr/lib/golang/src/github.com/containernetworking/cni/pkg/version (from $GOROOT)
        /home/user/go/src/github.com/containernetworking/cni/pkg/version (from $GOPATH)
make: *** [output/linux_amd64/cni/azure-vnet] Error 1
```

```
[user@localhost azure-container-networking]$ go get github.com/containernetworking/cni/pkg/invoke
[user@localhost  azure-container-networking]$ go get github.com/containernetworking/cni/pkg/skel
```

```
[user@localhost  azure-container-networking]$ make all-binaries
chmod 0755 output/linux_amd64/cnm/azure-vnet-plugin
cd output/linux_amd64/cnm && tar -czvf azure-vnet-cnm-linux-amd64-v1.0.5-1-g89555ac.tgz azure-vnet-plugin
azure-vnet-plugin
chown 1000:1000 output/linux_amd64/cnm/azure-vnet-cnm-linux-amd64-v1.0.5-1-g89555ac.tgz
go build -v -o output/linux_amd64/cni/azure-vnet -ldflags "-X main.version=v1.0.5-1-g89555ac -s -w" cni/network/plugin/*.go
github.com/Azure/azure-container-networking/platform
github.com/Azure/azure-container-networking/log
github.com/Azure/azure-container-networking/store
github.com/Azure/azure-container-networking/common
github.com/Azure/azure-container-networking/network/policy
github.com/Azure/azure-container-networking/cni
# github.com/Azure/azure-container-networking/cni
cni/plugin.go:112:39: not enough arguments in call to skel.PluginMainWithError
        have (func(*skel.CmdArgs) error, func(*skel.CmdArgs) error, version.PluginInfo)
        want (func(*skel.CmdArgs) error, func(*skel.CmdArgs) error, func(*skel.CmdArgs) error, version.PluginInfo)
github.com/Azure/azure-container-networking/ebtables
github.com/Azure/azure-container-networking/netlink
github.com/Azure/azure-container-networking/network
github.com/Azure/azure-container-networking/telemetry
make: *** [output/linux_amd64/cni/azure-vnet] Error 2
```

How to build this fucking shit?

I've tried:
```
[vagrant@localhost azure-container-networking]$ make all-binaries-containerized
pwd && ls -l
/home/vagrant/azure-container-networking
total 32
drwxrwxr-x.  2 vagrant vagrant   60 May 25 06:34 build
-rw-rw-r--.  1 vagrant vagrant 2379 May 25 06:34 CHANGELOG.md
drwxrwxr-x.  4 vagrant vagrant  161 May 25 06:34 cni
drwxrwxr-x.  5 vagrant vagrant  145 May 25 06:34 cnm
drwxrwxr-x. 10 vagrant vagrant  202 May 25 06:34 cns
drwxrwxr-x.  2 vagrant vagrant   90 May 25 06:34 common
-rw-rw-r--.  1 vagrant vagrant  323 May 25 06:34 Dockerfile.build
-rw-rw-r--.  1 vagrant vagrant  292 May 25 06:34 Dockerfile.cnm
drwxrwxr-x.  2 vagrant vagrant  116 May 25 06:34 docs
drwxrwxr-x.  2 vagrant vagrant   25 May 25 06:34 ebtables
drwxrwxr-x.  2 vagrant vagrant  121 May 25 06:34 ipam
-rw-rw-r--.  1 vagrant vagrant 1072 May 25 06:34 LICENSE
drwxrwxr-x.  2 vagrant vagrant  110 May 25 06:34 log
-rw-rw-r--.  1 vagrant vagrant  125 May 25 06:34 MAINTAINERS
-rw-rw-r--.  1 vagrant vagrant 6483 May 25 06:34 Makefile
drwxrwxr-x.  2 vagrant vagrant  111 May 25 06:34 netlink
drwxrwxr-x.  3 vagrant vagrant  211 May 25 06:34 network
drwxrwxr-x.  2 vagrant vagrant    6 May 25 06:35 output
drwxrwxr-x.  2 vagrant vagrant   59 May 25 06:34 platform
-rw-rw-r--.  1 vagrant vagrant 2544 May 25 06:34 README.md
drwxrwxr-x.  2 vagrant vagrant  112 May 25 06:34 scripts
drwxrwxr-x.  2 vagrant vagrant   57 May 25 06:34 store
drwxrwxr-x.  2 vagrant vagrant  105 May 25 06:34 telemetry
docker build -f Dockerfile.build -t acn-build:v1.0.5-1-g89555ac .
Sending build context to Docker daemon   5.17MB
Step 1/7 : FROM golang:1.8.1
 ---> fcef75ee6be1
Step 2/7 : RUN apt-get update && apt-get install -y zip
 ---> Using cache
 ---> ab03926180d6
Step 3/7 : RUN go get -d golang.org/x/sys/unix
 ---> Using cache
 ---> 6a2488338f85
Step 4/7 : RUN go get -d github.com/Microsoft/hcsshim
 ---> Using cache
 ---> bf7d69352c82
Step 5/7 : RUN go get -d github.com/containernetworking/cni/pkg/skel
 ---> Using cache
 ---> 46806fd56fdd
Step 6/7 : COPY . /go/src/github.com/Azure/azure-container-networking
 ---> Using cache
 ---> 3fa64a4b7430
Step 7/7 : WORKDIR /go/src/github.com/Azure/azure-container-networking
 ---> Using cache
 ---> ec0bd03ee3b4
Successfully built ec0bd03ee3b4
Successfully tagged acn-build:v1.0.5-1-g89555ac
docker run --name acn-builder \
        acn-build:v1.0.5-1-g89555ac \
        bash -c '\
                pwd && ls -l && \
                export GOOS=linux && \
                export GOARCH=amd64 && \
                make all-binaries && \
                chown -R 1000:1000 output/linux_amd64 \
        '
/go/src/github.com/Azure/azure-container-networking
total 32
-rw-rw-r--.  1 root root 2379 May 25 06:34 CHANGELOG.md
-rw-rw-r--.  1 root root  323 May 25 06:34 Dockerfile.build
-rw-rw-r--.  1 root root  292 May 25 06:34 Dockerfile.cnm
-rw-rw-r--.  1 root root 1072 May 25 06:34 LICENSE
-rw-rw-r--.  1 root root  125 May 25 06:34 MAINTAINERS
-rw-rw-r--.  1 root root 6483 May 25 06:34 Makefile
-rw-rw-r--.  1 root root 2544 May 25 06:34 README.md
drwxrwxr-x.  2 root root   60 May 25 06:34 build
drwxrwxr-x.  4 root root  161 May 25 06:34 cni
drwxrwxr-x.  5 root root  145 May 25 06:34 cnm
drwxrwxr-x. 10 root root  202 May 25 06:34 cns
drwxrwxr-x.  2 root root   90 May 25 06:34 common
drwxrwxr-x.  2 root root  116 May 25 06:34 docs
drwxrwxr-x.  2 root root   25 May 25 06:34 ebtables
drwxrwxr-x.  2 root root  121 May 25 06:34 ipam
drwxrwxr-x.  2 root root  110 May 25 06:34 log
drwxrwxr-x.  2 root root  111 May 25 06:34 netlink
drwxrwxr-x.  3 root root  211 May 25 06:34 network
drwxrwxr-x.  2 root root    6 May 25 06:35 output
drwxrwxr-x.  2 root root   59 May 25 06:34 platform
drwxrwxr-x.  2 root root  112 May 25 06:34 scripts
drwxrwxr-x.  2 root root   57 May 25 06:34 store
drwxrwxr-x.  2 root root  105 May 25 06:34 telemetry
go build -v -o output/linux_amd64/cnm/azure-vnet-plugin -ldflags "-X main.version=v1.0.5-1-g89555ac -s -w" cnm/plugin/*.go
github.com/Azure/azure-container-networking/platform
github.com/Azure/azure-container-networking/log
github.com/Azure/azure-container-networking/store
github.com/Azure/azure-container-networking/common
github.com/Azure/azure-container-networking/cnm
github.com/Azure/azure-container-networking/ipam
github.com/Azure/azure-container-networking/cnm/ipam
github.com/Azure/azure-container-networking/ebtables
golang.org/x/sys/unix
# golang.org/x/sys/unix
/go/src/golang.org/x/sys/unix/ztypes_linux_amd64.go:1405: syntax error: unexpected = in type declaration
/go/src/golang.org/x/sys/unix/ztypes_linux_amd64.go:1407: syntax error: non-declaration statement outside function body
github.com/Azure/azure-container-networking/network/policy
Makefile:102: recipe for target 'output/linux_amd64/cnm/azure-vnet-plugin' failed
make: *** [output/linux_amd64/cnm/azure-vnet-plugin] Error 2
make: *** [all-binaries-containerized] Error 2

```

--- a/cni/cni.go
+++ b/cni/cni.go
@@ -11,6 +11,7 @@ const (
 	// CNI commands.
 	Cmd    = "CNI_COMMAND"
 	CmdAdd = "ADD"
+	CmdGet = "GET"
 	CmdDel = "DEL"
 
 	// CNI errors.
@@ -21,10 +22,11 @@ const (
 )
 
 // Supported CNI versions.
-var supportedVersions = []string{"0.1.0", "0.2.0", "0.3.0", "0.3.1"}
+var supportedVersions = []string{"0.1.0", "0.2.0", "0.3.0", "0.3.1", "0.4.0"}
 
 // CNI contract.
 type PluginApi interface {
 	Add(args *cniSkel.CmdArgs) error
+	Get(args *cniSkel.CmdArgs) error
 	Delete(args *cniSkel.CmdArgs) error
 }

--- a/cni/ipam/ipam.go
+++ b/cni/ipam/ipam.go
@@ -250,6 +250,11 @@ func (plugin *ipamPlugin) Add(args *cniSkel.CmdArgs) error {
 	return nil
 }
 
+// Get handles CNI Get commands.
+func (plugin *ipamPlugin) Get(args *cniSkel.CmdArgs) error {
+	return nil
+}
+
 // Delete handles CNI delete commands.
 func (plugin *ipamPlugin) Delete(args *cniSkel.CmdArgs) error {
 	var err error

--- a/cni/network/network.go
+++ b/cni/network/network.go
@@ -23,12 +23,9 @@ import (
 const (
 	// Plugin name.
 	name = "azure-vnet"
-)
 
-const (
-	// Supported IP version
-	// Currently support only IPv4
-	ipVersion string = "4"
+	// Supported IP version. Currently support only IPv4
+	ipVersion = "4"
 )
 
 // NetPlugin represents the CNI network plugin.
@@ -434,8 +431,12 @@ func (plugin *netPlugin) Get(args *cniSkel.CmdArgs) error {
 			Version:   ipVersion,
 			Interface: &epInfo.IfIndex,
 			Address:   ipAddresses,
-			Gateway:   epInfo.Gateways[0],
 		}
+
+		if epInfo.Gateways != nil {
+			ipConfig.Gateway = epInfo.Gateways[0]
+		}
+
 		result.IPs = append(result.IPs, ipConfig)
 	}
 

--- a/cni/network/network.go
+++ b/cni/network/network.go
@@ -16,12 +16,19 @@ import (
 	"github.com/Azure/azure-container-networking/telemetry"
 
 	cniSkel "github.com/containernetworking/cni/pkg/skel"
+	cniTypes "github.com/containernetworking/cni/pkg/types"
 	cniTypesCurr "github.com/containernetworking/cni/pkg/types/current"
 )
 
 const (
 	// Plugin name.
 	name = "azure-vnet"
+)
+
+const (
+	// Supported IP version
+	// Currently support only IPv4
+	ipVersion string = "4"
 )
 
 // NetPlugin represents the CNI network plugin.
@@ -358,6 +365,86 @@ func (plugin *netPlugin) Add(args *cniSkel.CmdArgs) error {
 		err = plugin.Errorf("Failed to create endpoint: %v", err)
 		return err
 	}
+
+	return nil
+}
+
+// Get handles CNI Get commands.
+func (plugin *netPlugin) Get(args *cniSkel.CmdArgs) error {
+	var (
+		result cniTypesCurr.Result
+		err    error
+		nwCfg  *cni.NetworkConfig
+		epInfo *network.EndpointInfo
+		iface  *cniTypesCurr.Interface
+	)
+
+	log.Printf("[cni-net] Processing GET command with args {ContainerID:%v Netns:%v IfName:%v Args:%v Path:%v}.",
+		args.ContainerID, args.Netns, args.IfName, args.Args, args.Path)
+
+	defer func() {
+		// Add Interfaces to result.
+		iface = &cniTypesCurr.Interface{
+			Name: args.IfName,
+		}
+		result.Interfaces = append(result.Interfaces, iface)
+
+		if err == nil {
+			// Convert result to the requested CNI version.
+			res, err := result.GetAsVersion(nwCfg.CNIVersion)
+			if err != nil {
+				err = plugin.Error(err)
+			}
+			// Output the result to stdout.
+			res.Print()
+		}
+
+		log.Printf("[cni-net] GET command completed with result:%+v err:%v.", result, err)
+	}()
+
+	// Parse network configuration from stdin.
+	nwCfg, err = cni.ParseNetworkConfig(args.StdinData)
+	if err != nil {
+		err = plugin.Errorf("Failed to parse network configuration: %v.", err)
+		return err
+	}
+
+	log.Printf("[cni-net] Read network configuration %+v.", nwCfg)
+
+	// Initialize values from network config.
+	networkId := nwCfg.Name
+	endpointId := GetEndpointID(args)
+
+	// Query the network.
+	_, err = plugin.nm.GetNetworkInfo(networkId)
+	if err != nil {
+		plugin.Errorf("Failed to query network: %v", err)
+		return err
+	}
+
+	// Query the endpoint.
+	epInfo, err = plugin.nm.GetEndpointInfo(networkId, endpointId)
+	if err != nil {
+		plugin.Errorf("Failed to query endpoint: %v", err)
+		return err
+	}
+
+	for _, ipAddresses := range epInfo.IPAddresses {
+		ipConfig := &cniTypesCurr.IPConfig{
+			Version:   ipVersion,
+			Interface: &epInfo.IfIndex,
+			Address:   ipAddresses,
+			Gateway:   epInfo.Gateways[0],
+		}
+		result.IPs = append(result.IPs, ipConfig)
+	}
+
+	for _, route := range epInfo.Routes {
+		result.Routes = append(result.Routes, &cniTypes.Route{Dst: route.Dst, GW: route.Gw})
+	}
+
+	result.DNS.Nameservers = epInfo.DNS.Servers
+	result.DNS.Domain = epInfo.DNS.Suffix
 
 	return nil
 }

--- a/cni/network/plugin/main.go
+++ b/cni/network/plugin/main.go
@@ -83,11 +83,30 @@ func main() {
 
 	netPlugin.SetReportManager(reportManager)
 
+	defer func() {
+		errUninit := netPlugin.Plugin.UninitializeKeyValueStore()
+		if errUninit != nil {
+			log.Printf("Failed to uninitialize key-value store of network plugin, err:%v.\n", err)
+		}
+
+		if recover() != nil {
+			os.Exit(1)
+		}
+	}()
+
+	err = netPlugin.Plugin.InitializeKeyValueStore(&config)
+
+	if err != nil {
+		log.Printf("Failed to initialize key-value store of network plugin, err:%v.\n", err)
+		reportPluginError(reportManager, err)
+		panic("network plugin fatal error")
+	}
+
 	err = netPlugin.Start(&config)
 	if err != nil {
 		log.Printf("Failed to start network plugin, err:%v.\n", err)
 		reportPluginError(reportManager, err)
-		os.Exit(1)
+		panic("network plugin fatal error")
 	}
 
 	err = netPlugin.Execute(cni.PluginApi(netPlugin))
@@ -99,7 +118,7 @@ func main() {
 	netPlugin.Stop()
 
 	if err != nil {
-		os.Exit(1)
+		panic("network plugin fatal error")
 	}
 
 	// Report CNI successfully finished execution.

--- a/cni/network/plugin/main.go
+++ b/cni/network/plugin/main.go
@@ -84,8 +84,7 @@ func main() {
 	netPlugin.SetReportManager(reportManager)
 
 	defer func() {
-		errUninit := netPlugin.Plugin.UninitializeKeyValueStore()
-		if errUninit != nil {
+		if errUninit := netPlugin.Plugin.UninitializeKeyValueStore(); errUninit != nil {
 			log.Printf("Failed to uninitialize key-value store of network plugin, err:%v.\n", err)
 		}
 
@@ -94,23 +93,19 @@ func main() {
 		}
 	}()
 
-	err = netPlugin.Plugin.InitializeKeyValueStore(&config)
-
-	if err != nil {
+	if err = netPlugin.Plugin.InitializeKeyValueStore(&config); err != nil {
 		log.Printf("Failed to initialize key-value store of network plugin, err:%v.\n", err)
 		reportPluginError(reportManager, err)
 		panic("network plugin fatal error")
 	}
 
-	err = netPlugin.Start(&config)
-	if err != nil {
+	if err = netPlugin.Start(&config); err != nil {
 		log.Printf("Failed to start network plugin, err:%v.\n", err)
 		reportPluginError(reportManager, err)
 		panic("network plugin fatal error")
 	}
 
-	err = netPlugin.Execute(cni.PluginApi(netPlugin))
-	if err != nil {
+	if err = netPlugin.Execute(cni.PluginApi(netPlugin)); err != nil {
 		log.Printf("Failed to execute network plugin, err:%v.\n", err)
 		reportPluginError(reportManager, err)
 	}
@@ -123,8 +118,7 @@ func main() {
 
 	// Report CNI successfully finished execution.
 	reportManager.Report.CniSucceeded = true
-	err = reportManager.SendReport()
-	if err != nil {
+	if err = reportManager.SendReport(); err != nil {
 		log.Printf("SendReport failed due to %v", err)
 	} else {
 		markSendReport(reportManager)

--- a/cni/plugin.go
+++ b/cni/plugin.go
@@ -162,16 +162,16 @@ func (plugin *Plugin) InitializeKeyValueStore(config *common.PluginConfig) error
 			log.Printf("[cni] Failed to create store, err:%v.", err)
 			return err
 		}
-
-		// Acquire store lock.
-		err = plugin.Store.Lock(true)
-		if err != nil {
-			log.Printf("[cni] Timed out on locking store, err:%v.", err)
-			return err
-		}
-
-		config.Store = plugin.Store
 	}
+
+	// Acquire store lock.
+	if err := plugin.Store.Lock(true); err != nil {
+		log.Printf("[cni] Timed out on locking store, err:%v.", err)
+		return err
+	}
+
+	config.Store = plugin.Store
+
 	return nil
 }
 
@@ -185,5 +185,6 @@ func (plugin *Plugin) UninitializeKeyValueStore() error {
 		}
 	}
 	plugin.Store = nil
+
 	return nil
 }

--- a/cni/plugin.go
+++ b/cni/plugin.go
@@ -51,37 +51,11 @@ func (plugin *Plugin) Initialize(config *common.PluginConfig) error {
 		log.Printf("[cni] Failed to configure logging, err:%v.\n", err)
 		return err
 	}
-
-	// Initialize store.
-	if plugin.Store == nil {
-		// Create the key value store.
-		var err error
-		plugin.Store, err = store.NewJsonFileStore(platform.CNIRuntimePath + plugin.Name + ".json")
-		if err != nil {
-			log.Printf("[cni] Failed to create store, err:%v.", err)
-			return err
-		}
-
-		// Acquire store lock.
-		err = plugin.Store.Lock(true)
-		if err != nil {
-			log.Printf("[cni] Timed out on locking store, err:%v.", err)
-			return err
-		}
-
-		config.Store = plugin.Store
-	}
-
 	return nil
 }
 
 // Uninitialize uninitializes the plugin.
 func (plugin *Plugin) Uninitialize() {
-	err := plugin.Store.Unlock()
-	if err != nil {
-		log.Printf("[cni] Failed to unlock store, err:%v.", err)
-	}
-
 	plugin.Plugin.Uninitialize()
 }
 
@@ -109,7 +83,7 @@ func (plugin *Plugin) Execute(api PluginApi) (err error) {
 	pluginInfo := cniVers.PluginSupports(supportedVersions...)
 
 	// Parse args and call the appropriate cmd handler.
-	cniErr := cniSkel.PluginMainWithError(api.Add, api.Delete, pluginInfo)
+	cniErr := cniSkel.PluginMainWithError(api.Add, api.Get, api.Delete, pluginInfo)
 	if cniErr != nil {
 		cniErr.Print()
 		return cniErr
@@ -176,4 +150,40 @@ func (plugin *Plugin) Error(err error) *cniTypes.Error {
 // Errorf creates and logs a custom CNI error according to a format specifier.
 func (plugin *Plugin) Errorf(format string, args ...interface{}) *cniTypes.Error {
 	return plugin.Error(fmt.Errorf(format, args...))
+}
+
+// Initialize key-value store
+func (plugin *Plugin) InitializeKeyValueStore(config *common.PluginConfig) error {
+	// Create the key value store.
+	if plugin.Store == nil {
+		var err error
+		plugin.Store, err = store.NewJsonFileStore(platform.CNIRuntimePath + plugin.Name + ".json")
+		if err != nil {
+			log.Printf("[cni] Failed to create store, err:%v.", err)
+			return err
+		}
+
+		// Acquire store lock.
+		err = plugin.Store.Lock(true)
+		if err != nil {
+			log.Printf("[cni] Timed out on locking store, err:%v.", err)
+			return err
+		}
+
+		config.Store = plugin.Store
+	}
+	return nil
+}
+
+// Uninitialize key-value store
+func (plugin *Plugin) UninitializeKeyValueStore() error {
+	if plugin.Store != nil {
+		err := plugin.Store.Unlock()
+		if err != nil {
+			log.Printf("[cni] Failed to unlock store, err:%v.", err)
+			return err
+		}
+	}
+	plugin.Store = nil
+	return nil
 }

--- a/network/endpoint.go
+++ b/network/endpoint.go
@@ -21,6 +21,8 @@ type endpoint struct {
 	MacAddress  net.HardwareAddr
 	IPAddresses []net.IPNet
 	Gateways    []net.IP
+	DNS         DNSInfo
+	Routes      []RouteInfo
 }
 
 // EndpointInfo contains read-only information about an endpoint.
@@ -34,6 +36,10 @@ type EndpointInfo struct {
 	DNS         DNSInfo
 	Policies    []policy.Policy
 	Data        map[string]interface{}
+	MacAddress  net.HardwareAddr
+	SandboxKey  string
+	Gateways    []net.IP
+	IfIndex     int
 }
 
 // RouteInfo contains information about an IP route.
@@ -144,6 +150,18 @@ func (ep *endpoint) getInfo() *EndpointInfo {
 		Id:          ep.Id,
 		IPAddresses: ep.IPAddresses,
 		Data:        make(map[string]interface{}),
+		MacAddress:  ep.MacAddress,
+		SandboxKey:  ep.SandboxKey,
+		IfIndex:     0, // Azure CNI supports only one interface
+		DNS:         ep.DNS,
+	}
+
+	for _, route := range ep.Routes {
+		info.Routes = append(info.Routes, route)
+	}
+
+	for _, gw := range ep.Gateways {
+		info.Gateways = append(info.Gateways, gw)
 	}
 
 	// Call the platform implementation.

--- a/network/endpoint.go
+++ b/network/endpoint.go
@@ -31,15 +31,15 @@ type EndpointInfo struct {
 	ContainerID string
 	NetNsPath   string
 	IfName      string
+	SandboxKey  string
+	IfIndex     int
+	MacAddress  net.HardwareAddr
+	DNS         DNSInfo
 	IPAddresses []net.IPNet
 	Routes      []RouteInfo
-	DNS         DNSInfo
 	Policies    []policy.Policy
-	Data        map[string]interface{}
-	MacAddress  net.HardwareAddr
-	SandboxKey  string
 	Gateways    []net.IP
-	IfIndex     int
+	Data        map[string]interface{}
 }
 
 // RouteInfo contains information about an IP route.

--- a/network/endpoint_linux.go
+++ b/network/endpoint_linux.go
@@ -222,6 +222,11 @@ func (nw *network) newEndpointImpl(epInfo *EndpointInfo) (*endpoint, error) {
 		MacAddress:  containerIf.HardwareAddr,
 		IPAddresses: epInfo.IPAddresses,
 		Gateways:    []net.IP{nw.extIf.IPv4Gateway},
+		DNS:         epInfo.DNS,
+	}
+
+	for _, route := range epInfo.Routes {
+		ep.Routes = append(ep.Routes, route)
 	}
 
 	return ep, nil

--- a/network/endpoint_windows.go
+++ b/network/endpoint_windows.go
@@ -70,6 +70,11 @@ func (nw *network) newEndpointImpl(epInfo *EndpointInfo) (*endpoint, error) {
 		IfName:      epInfo.IfName,
 		IPAddresses: epInfo.IPAddresses,
 		Gateways:    []net.IP{net.ParseIP(hnsResponse.GatewayAddress)},
+		DNS:         epInfo.DNS,
+	}
+
+	for _, route := range epInfo.Routes {
+		ep.Routes = append(ep.Routes, route)
 	}
 
 	ep.MacAddress, _ = net.ParseMAC(hnsResponse.MacAddress)

